# Indeed Data Scraping Project
The goal of this project is to automate Indeed Job searching by allowing the user to input a state of their choice and returning a CSV with the job listings in the first 10 pages on Indeed. This will the job searching process much easier for the user.

## Import Statements

In [1]:
from bs4 import BeautifulSoup as BSoup
import requests
import pandas as pd

## Let's first start by attempting to create a dataframe from just one Indeed page URL. Note this is a sample URL.

In [2]:
URL = "https://www.indeed.com/q-Data-Scientist-l-San-Francisco,-CA-jobs.html?vjk=bc7c0e642f6453f4"
request = requests.get(URL)
print(request)

<Response [200]>


#### Awesome, we got a response code 200 meaning our request was successful! Let's now view the page as HTML and use BeautifulSoup to make it look nicer. I will comment out the print statement so it won't display the whole html because it is very long :)

In [3]:
page_html = BSoup(request.text, "html.parser")
#print(page_html.prettify())

In [4]:
containers = page_html.findAll(name="div", attrs={"class": "row"})
len(containers)

15

#### It looks like there are 15 job from this sample URL. Let's now try to extract the job title from each listing by looking at the HTML tags from the page_html variable.

In [5]:
def extract_job_title_from_result(soup): 
    jobs = []
    for container in containers:
        for a in container.find_all(name="a", attrs={"data-tn-element": "jobTitle"}):
            jobs.append(a["title"])
    return jobs

extract_job_title_from_result(page_html)

['Data Scientist',
 'Data Scientist',
 'Data Scientist',
 'Associate Data Scientist I',
 'Data Scientist',
 'Data Scientist: Data Visualization',
 'Research Data Scientist',
 'Junior Data Scientist - Pathrise Recruiting Partners',
 'Data Scientist',
 'Data Scientist',
 'Data Scientist',
 'Data Scientist / Quantitative Research',
 'Data Scientist',
 'Data Scientist, Marketing',
 'Data Scientist, Legal Policy & Economics']

#### Let's do the same for the company.

In [6]:
def extract_company_from_result(soup): 
    companies = []
    for container in containers:
        company = container.find_all(name="span", attrs={"class": "company"})
        if len(company) > 0:
            for b in company:
                companies.append(b.text.strip())
        else:
            test2 = div.find_all(name="span", attrs={"class": "result-link-source"})
            for span in test2:
                companies.append(span.text.strip())
    return(companies)
 
extract_company_from_result(page_html)

['Triplebyte',
 'Global Fishing Watch',
 'Blue Owl',
 'Levi Strauss & Co.',
 'project AI',
 'Kaiser Permanente',
 'University of California San Francisco',
 'Pathrise',
 'Common Networks',
 'Applied Technology & Science (A-T-S)',
 'Yelp',
 'PicnicHealth',
 'GradTests (gradtests.com)',
 'Twitch',
 'Uber']

#### Let's do the same for the salary.

In [7]:
def extract_salary_from_result(soup): 
    salaries = []
    for div in soup.find_all(name="div", attrs={"class": "row"}):
        div_two = div.find(name="span", attrs={'class': "salaryText"})
        if div_two == None:
            salaries.append("Not Available")
        else:
            salaries.append(div_two.text.strip())
    return salaries 

extract_salary_from_result(page_html)

['$145,000 - $225,000 a year',
 '$45 - $65 an hour',
 '$250,000 - $375,000 a year',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 'Not Available',
 '$120,000 a year',
 'Not Available',
 'Not Available']

#### And finally, let's do the same for ratings.

In [8]:
def extract_ratings_from_result(soup): 
    ratings = []
    for div in soup.find_all(name="div", attrs={"class": "row"}):
        div_two = div.find(name="span", attrs={'class': "ratingsContent"})
        if div_two == None:
            ratings.append("Not Available")
        else:
            ratings.append(div_two.text.strip())
    return ratings

extract_ratings_from_result(page_html)

['5.0',
 'Not Available',
 'Not Available',
 '3.9',
 'Not Available',
 '4.1',
 '4.2',
 'Not Available',
 'Not Available',
 'Not Available',
 '3.5',
 'Not Available',
 'Not Available',
 '4.4',
 '3.7']

#### Now let's build a dataframe by combining all the information we have so far!

In [9]:
example_df = pd.DataFrame({"job_title": extract_job_title_from_result(page_html), 
                         "company": extract_company_from_result(page_html),
                         "salary": extract_salary_from_result(page_html),
                         "rating": extract_ratings_from_result(page_html)}) 
example_df

job_title  \
0                                      Data Scientist   
1                                      Data Scientist   
2                                      Data Scientist   
3                          Associate Data Scientist I   
4                                      Data Scientist   
5                  Data Scientist: Data Visualization   
6                             Research Data Scientist   
7   Junior Data Scientist - Pathrise Recruiting Pa...   
8                                      Data Scientist   
9                                      Data Scientist   
10                                     Data Scientist   
11             Data Scientist / Quantitative Research   
12                                     Data Scientist   
13                          Data Scientist, Marketing   
14           Data Scientist, Legal Policy & Economics   

                                   company                      salary  \
0                               Triplebyte  $145,000 - $225,000 a year   
1                     Global Fishing Watch           $45 - $65 an hour   
2                                 Blue Owl  $250,000 - $375,000 a year   
3                       Levi Strauss & Co.               Not Available   
4                               project AI               Not Available   
5                        Kaiser Permanente               Not Available   
6   University of California San Francisco               Not Available   
7                                 Pathrise               Not Available   
8                          Common Networks               Not Available   
9     Applied Technology & Science (A-T-S)               Not Available   
10                                    Yelp               Not Available   
11                            PicnicHealth               Not Available   
12               GradTests (gradtests.com)             $120,000 a year   
13                                  Twitch               Not Available   
14                                    Uber               Not Available   

           rating  
0             5.0  
1   Not Available  
2   Not Available  
3             3.9  
4   Not Available  
5             4.1  
6             4.2  
7   Not Available  
8   Not Available  
9   Not Available  
10            3.5  
11  Not Available  
12  Not Available  
13            4.4  
14            3.7

Awesome! It looks good except that we need to clean the salary series since it is not consistent with units (years and hour) and fix the style of the text. We won't worry too much about that right now. Let's now try to get all listings from the first 10 pages of the Indeed searches

In [10]:
limit = 100
columns = ["job_title", "company", "salary", "rating"]

In [11]:
a = input() # We want the user to be able to input a city of their choice, so we will test it using the input method
city_selection = [a]

San Francisco


In [12]:
#Create a df
sample_df = pd.DataFrame(columns=columns)

#loops through the 10 url for the selected city and gets the information
for city in city_selection:
    for start in range(0, limit, 10):
        page = requests.get("http://www.indeed.com/jobs?q=data+scientist+%2420%2C000&l=" + str(city) + "&start=" + str(start))
        soup = BSoup(page.text, "lxml", from_encoding="utf-8")
        
        for div in soup.find_all(name="div", attrs={"class": "row"}): 
            num = (len(sample_df) + 1) 
            job_post = [] 
            
            for a in div.find_all(name="a", attrs={"data-tn-element": "jobTitle"}):
                job_post.append(a["title"])
            company = div.find_all(name="span", attrs={"class": "company"}) 
            if len(company) > 0: 
                for b in company:
                    job_post.append(b.text.strip()) 
            else: 
                test2 = div.find_all(name="span", attrs={"class": "result-link-source"})
                for span in test2:
                    job_post.append(span.text)
                    
            div_two = div.find(name="span", attrs={"class": "salaryText"})
            if div_two == None:
                job_post.append("Not Available")
            else:
                job_post.append(div_two.text.strip())
                
            div_three = div.find(name="span", attrs={"class": "ratingsContent"})
            if div_three == None:
                job_post.append("Not Available")
            else:
                job_post.append(div_three.text.strip())
            
            sample_df.loc[num] = job_post

/srv/app/venv/lib/python3.6/site-packages/bs4/__init__.py:146: UserWarning: You provided Unicode markup but also provided a value for from_encoding. Your from_encoding will be ignored.
  warnings.warn("You provided Unicode markup but also provided a value for from_encoding. Your from_encoding will be ignored.")


In [13]:
sample_df

job_title  \
1                                       Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                           Associate Data Scientist I   
5                                       Data Scientist   
6                              Research Data Scientist   
7    Junior Data Scientist - Pathrise Recruiting Pa...   
8                   Data Scientist: Data Visualization   
9                                       Data Scientist   
10                                      Data Scientist   
11                                      Data Scientist   
12              Data Scientist / Quantitative Research   
13                                      Data Scientist   
14            Data Scientist, Legal Policy & Economics   
15                           Data Scientist, Marketing   
16                                      Data Scientist   
17                                      Data Scientist   
18            Data Scientist, Legal Policy & Economics   
19                           Data Scientist, Marketing   
20          Data Scientist, Machine Learning innovator   
21   Data Scientist – Experimentation (Contract Pos...   
22               Data Scientist - Shop Recommendations   
23                        Data Science Intern - Remote   
24                                      Data Scientist   
25                                      Data scientist   
26                               Senior Data Scientist   
27   Data Solutions Sr. Consultant/Personalization ...   
28                                      Data Scientist   
29                            Senior Applied Scientist   
30                                      Data Scientist   
..                                                 ...   
125                     Scheduling/Cost Analyst Intern   
126                           Director of Data Science   
127                Lead Data Scientist - Deep Learning   
128                              Senior Data Scientist   
129                               Chief Data Scientist   
130  Staff Data Scientist – Global Payments & Fraud...   
131                   Senior Data Scientist, Education   
132                                 Sr. Data Scientist   
133                                Senior Data Analyst   
134                                      Data Engineer   
135                         Senior Health Data Analyst   
136  Machine Learning Researcher - Multi View & Seg...   
137    Machine Learning Modeler Lead, Financial Crimes   
138                       Data Science Manager- Health   
139                                     Data Scientist   
140                                     Data Scientist   
141                              Senior Data Scientist   
142                              Senior Data Scientist   
143                                Senior Data Analyst   
144                              Senior Data Scientist   
145                  Backend Engineer/Machine Learning   
146                              Expert Data Scientist   
147                          Data Science/Stats Intern   
148                       Data Scientist (6257U) - CED   
149                          Machine Learning Engineer   
150  Financial Analyst Summer Intern - Ideal for a ...   
151                           Senior Applied Scientist   
152                                     Data Scientist   
153                                     Data Scientist   
154                                     Data Scientist   

                                        company                      salary  \
1                                    Triplebyte  $145,000 - $225,000 a year   
2                          Global Fishing Watch           $45 - $65 an hour   
3                                      Blue Owl  $250,000 - $375,000 a year   
4                            Levi Strauss & Co.               Not Available   
5                                    project AI      

#### Let's first fix the salary series since there is a bolding issue with the concatenation of the '$', the first number of the range, and '-'

In [14]:
lst = []

for i in range(1, len(sample_df['salary']) + 1):
    # 6 figure first range num for year
    if len(sample_df['salary'][i]) == 26:
        lst.append(sample_df['salary'][i][0:8] + ' - ' + sample_df['salary'][i][12:19] + ' a year')
    # 5 figure first range num for year
    elif len(sample_df['salary'][i]) == 25:
        lst.append(sample_df['salary'][i][0:7] + ' - ' + sample_df['salary'][i][11:18] + ' a year')
    # 6 figure for year no range
    elif len(sample_df['salary'][i]) == 15:
        lst.append(sample_df['salary'][i][0:8] + ' a year')  
    # 2 figure for hours
    elif len(sample_df['salary'][i]) == 17:
        lst.append(sample_df['salary'][i][0:3] + ' - ' + sample_df['salary'][i][7:9] + ' an hour')
    # one 3 figure for hours
    elif len(sample_df['salary'][i]) == 18:
        lst.append(sample_df['salary'][i][0:3] + ' - ' + sample_df['salary'][i][7:9] + ' an hour')
    # 4 figure for months
    elif len(sample_df['salary'][i]) == 14:
        lst.append(sample_df['salary'][i][0:6] + ' a month')
    else:
        lst.append(sample_df['salary'][i])
        
sample_df['salary'] = lst

sample_df

job_title  \
1                                       Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                           Associate Data Scientist I   
5                                       Data Scientist   
6                              Research Data Scientist   
7    Junior Data Scientist - Pathrise Recruiting Pa...   
8                   Data Scientist: Data Visualization   
9                                       Data Scientist   
10                                      Data Scientist   
11                                      Data Scientist   
12              Data Scientist / Quantitative Research   
13                                      Data Scientist   
14            Data Scientist, Legal Policy & Economics   
15                           Data Scientist, Marketing   
16                                      Data Scientist   
17                                      Data Scientist   
18            Data Scientist, Legal Policy & Economics   
19                           Data Scientist, Marketing   
20          Data Scientist, Machine Learning innovator   
21   Data Scientist – Experimentation (Contract Pos...   
22               Data Scientist - Shop Recommendations   
23                        Data Science Intern - Remote   
24                                      Data Scientist   
25                                      Data scientist   
26                               Senior Data Scientist   
27   Data Solutions Sr. Consultant/Personalization ...   
28                                      Data Scientist   
29                            Senior Applied Scientist   
30                                      Data Scientist   
..                                                 ...   
125                     Scheduling/Cost Analyst Intern   
126                           Director of Data Science   
127                Lead Data Scientist - Deep Learning   
128                              Senior Data Scientist   
129                               Chief Data Scientist   
130  Staff Data Scientist – Global Payments & Fraud...   
131                   Senior Data Scientist, Education   
132                                 Sr. Data Scientist   
133                                Senior Data Analyst   
134                                      Data Engineer   
135                         Senior Health Data Analyst   
136  Machine Learning Researcher - Multi View & Seg...   
137    Machine Learning Modeler Lead, Financial Crimes   
138                       Data Science Manager- Health   
139                                     Data Scientist   
140                                     Data Scientist   
141                              Senior Data Scientist   
142                              Senior Data Scientist   
143                                Senior Data Analyst   
144                              Senior Data Scientist   
145                  Backend Engineer/Machine Learning   
146                              Expert Data Scientist   
147                          Data Science/Stats Intern   
148                       Data Scientist (6257U) - CED   
149                          Machine Learning Engineer   
150  Financial Analyst Summer Intern - Ideal for a ...   
151                           Senior Applied Scientist   
152                                     Data Scientist   
153                                     Data Scientist   
154                                     Data Scientist   

                                        company                     salary  \
1                                    Triplebyte  $145,000 - 225,000 a year   
2                          Global Fishing Watch           $45 - 65 an hour   
3                                      Blue Owl  $250,000 - 375,000 a year   
4                            Levi Strauss & Co.              Not Available   
5                                    project AI           

#### Next, we are going to clean this data and then convert the dataframe into a CSV file. Let's start by cleaning the salary series to the correct rates. We will convert them into dollars a year. 

##### Note: the string manipulation below assumes that hourly salaries are two digits and monthly to be in the thousands since salaries for these jobs. We can safely make this assumption for now as annual income for this position is usually 50k-200k which supports that range.

In [15]:
result = sample_df['salary']

for index, item in enumerate(sample_df["salary"]):
    if "hour" in item and '-' in item:
        lower = int(item[1:3])*8*365
        upper = int(item[6:8])*8*365
        result[index + 1] = "$" + "{:,}".format(lower) + " - " + "{:,}".format(upper) + " a year"
    elif "hour" in item and '-' not in item:
        salary = int(item[1:3])*8*365
        result[index + 1] = "$" + "{:,}".format(salary) + " a year"
    elif 'month' in item:
        no_range = int(item[1:2] + item[3:6])*12
        result[index + 1] = "$" + "{:,}".format(no_range) + " a year"
    else:
        result[index + 1] = item
        
result

1      $145,000 - 225,000 a year
2      $131,400 - 189,800 a year
3      $250,000 - 375,000 a year
4                  Not Available
5                  Not Available
6                  Not Available
7                  Not Available
8                  Not Available
9                  Not Available
10                 Not Available
11                 Not Available
12                 Not Available
13               $120,000 a year
14                 Not Available
15                 Not Available
16     $145,000 - 225,000 a year
17     $131,400 - 189,800 a year
18                 Not Available
19                 Not Available
20                 Not Available
21                 Not Available
22                 Not Available
23                 Not Available
24                 Not Available
25                 Not Available
26                 Not Available
27                 Not Available
28     $250,000 - 375,000 a year
29                 Not Available
30     $131,400 - 189,800 a year
          

In [16]:
sample_df

job_title  \
1                                       Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                           Associate Data Scientist I   
5                                       Data Scientist   
6                              Research Data Scientist   
7    Junior Data Scientist - Pathrise Recruiting Pa...   
8                   Data Scientist: Data Visualization   
9                                       Data Scientist   
10                                      Data Scientist   
11                                      Data Scientist   
12              Data Scientist / Quantitative Research   
13                                      Data Scientist   
14            Data Scientist, Legal Policy & Economics   
15                           Data Scientist, Marketing   
16                                      Data Scientist   
17                                      Data Scientist   
18            Data Scientist, Legal Policy & Economics   
19                           Data Scientist, Marketing   
20          Data Scientist, Machine Learning innovator   
21   Data Scientist – Experimentation (Contract Pos...   
22               Data Scientist - Shop Recommendations   
23                        Data Science Intern - Remote   
24                                      Data Scientist   
25                                      Data scientist   
26                               Senior Data Scientist   
27   Data Solutions Sr. Consultant/Personalization ...   
28                                      Data Scientist   
29                            Senior Applied Scientist   
30                                      Data Scientist   
..                                                 ...   
125                     Scheduling/Cost Analyst Intern   
126                           Director of Data Science   
127                Lead Data Scientist - Deep Learning   
128                              Senior Data Scientist   
129                               Chief Data Scientist   
130  Staff Data Scientist – Global Payments & Fraud...   
131                   Senior Data Scientist, Education   
132                                 Sr. Data Scientist   
133                                Senior Data Analyst   
134                                      Data Engineer   
135                         Senior Health Data Analyst   
136  Machine Learning Researcher - Multi View & Seg...   
137    Machine Learning Modeler Lead, Financial Crimes   
138                       Data Science Manager- Health   
139                                     Data Scientist   
140                                     Data Scientist   
141                              Senior Data Scientist   
142                              Senior Data Scientist   
143                                Senior Data Analyst   
144                              Senior Data Scientist   
145                  Backend Engineer/Machine Learning   
146                              Expert Data Scientist   
147                          Data Science/Stats Intern   
148                       Data Scientist (6257U) - CED   
149                          Machine Learning Engineer   
150  Financial Analyst Summer Intern - Ideal for a ...   
151                           Senior Applied Scientist   
152                                     Data Scientist   
153                                     Data Scientist   
154                                     Data Scientist   

                                        company                     salary  \
1                                    Triplebyte  $145,000 - 225,000 a year   
2                          Global Fishing Watch  $131,400 - 189,800 a year   
3                                      Blue Owl  $250,000 - 375,000 a year   
4                            Levi Strauss & Co.              Not Available   
5                                    project AI           

#### Awesome, it looks like it worked properly. For example, in our old series for index 134, it was 5,000 monthly which converts to 60,000 a year in our results series. Let's now finish off by printing our whole dataframe out.

In [17]:
pd.set_option('display.max_rows', None)
sample_df

job_title  \
1                                       Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                           Associate Data Scientist I   
5                                       Data Scientist   
6                              Research Data Scientist   
7    Junior Data Scientist - Pathrise Recruiting Pa...   
8                   Data Scientist: Data Visualization   
9                                       Data Scientist   
10                                      Data Scientist   
11                                      Data Scientist   
12              Data Scientist / Quantitative Research   
13                                      Data Scientist   
14            Data Scientist, Legal Policy & Economics   
15                           Data Scientist, Marketing   
16                                      Data Scientist   
17                                      Data Scientist   
18            Data Scientist, Legal Policy & Economics   
19                           Data Scientist, Marketing   
20          Data Scientist, Machine Learning innovator   
21   Data Scientist – Experimentation (Contract Pos...   
22               Data Scientist - Shop Recommendations   
23                        Data Science Intern - Remote   
24                                      Data Scientist   
25                                      Data scientist   
26                               Senior Data Scientist   
27   Data Solutions Sr. Consultant/Personalization ...   
28                                      Data Scientist   
29                            Senior Applied Scientist   
30                                      Data Scientist   
31                                      Data Scientist   
32                            Senior Applied Scientist   
33                                      Data Scientist   
34                                      Data Scientist   
35                                      Data Scientist   
36                                Staff Data Scientist   
37                                      Data Scientist   
38   2020 PhD University Graduate - Data Scientist ...   
39                                      Data Scientist   
40                                      Data Scientist   
41                                      Data Scientist   
42                           Data Scientist, Analytics   
43             Enterprise Architect (Data) (2020-9124)   
44   Financial Analyst Summer Intern - Ideal for a ...   
45                                      Data Scientist   
46                                      Data Scientist   
47   Financial Analyst Summer Intern - Ideal for a ...   
48                            Senior Applied Scientist   
49                         Data Scientist, Forecasting   
50                                      Data Scientist   
51              Data Scientist, Transaction Monitoring   
52                Data Scientist/Analyst - Wilbur Labs   
53                         Data Scientist - Operations   
54                                      Data Scientist   
55                                      Data Scientist   
56                                      Data Scientist   
57                  Data Scientist - Product Analytics   
58                                      Data Scientist   
59                           Machine Learning Engineer   
60                                      Data Scientist   
61                                      Data Scientist   
62                                 Data Science Intern   
63                                      Data Scientist   
64          Sr. Data Scientist, Marketing Data Quality   
65   2020 PhD Data Scientist Internship - Uber Eats...   
66               Sr. Data Scientist, Consumer Behavior   
67                     Data Scientist, Creator Content   
68                      Data Scientist, Trust & Safety   
69              Senior Data Scientist, Ad I

#### Looks good! Let's now finally convert this Pandas dataframe into a CSV file.

In [18]:
sample_df.to_csv('indeed.csv', index=False)